src/Swarm/Game/Step.hs

```diff
@@ -132,39 +132,123 @@ gameTick = do
+--    that prerequisites are evaluated before dependent goals.
+-- 3) The iteration needs to be a "fold", so that state is updated
+--    after each element.
+hypotheticalWinCheck ::
+  (Has (State GameState) sig m, Has (Lift IO) sig m) =>
+  EntityMap ->
+  GameState ->
+  WinStatus ->
+  ObjectiveCompletion ->
+  m ()
+hypotheticalWinCheck em g ws oc = do
+  -- We can fully and accurately evaluate the new state of the objectives DAG
+  -- in a single pass, so long as we visit it in reverse topological order.
+  --
+  -- N.B. The "reverse" is essential due to the re-population of the
+  -- "incomplete" goal list by cons-ing.
+  finalAccumulator <-
+    foldM foldFunc initialAccumulator $
+      reverse incompleteGoals
+
+  let newWinState = case ws of
+        Ongoing -> getNextWinState $ completions finalAccumulator
+        _ -> ws
+
+  winCondition .= WinConditions newWinState (completions finalAccumulator)
+  announcementQueue %= (>< Seq.fromList (map ObjectiveCompleted $ completionAnnouncementQueue finalAccumulator))
+
+  mapM_ handleException $ exceptions finalAccumulator
+ where
+  getNextWinState comps
+    | WC.didWin comps = Won False
+    | WC.didLose comps = Unwinnable False
+    | otherwise = Ongoing
+
+  (withoutIncomplete, incompleteGoals) = OB.extractIncomplete oc
+  initialAccumulator = CompletionsWithExceptions [] withoutIncomplete []
+
+  -- All of the "incomplete" goals have been emptied from the initial accumulator, and
+  -- these are what we iterate over with the fold.
+  -- Each iteration, we either place the goal back into the "incomplete" bucket, or
+  -- we determine that it has been met and place it into the "completed" bucket.
+  foldFunc (CompletionsWithExceptions exnTexts currentCompletions announcements) obj = do
+    v <-
+      if WC.isPrereqsSatisfied currentCompletions obj
+        then runThrow @Exn . evalState @GameState g $ evalPT $ obj ^. OB.objectiveCondition
+        else return $ Right $ VBool False
+    let simplified = simplifyResult v
+    return $ case simplified of
+      Left exnText ->
+        CompletionsWithExceptions
+          (exnText : exnTexts)
+          currentCompletions
+          announcements
+      Right boolResult ->
+        CompletionsWithExceptions
+          exnTexts
+          (modifyCompletions obj currentCompletions)
+          (modifyAnnouncements announcements)
+       where
+        (modifyCompletions, modifyAnnouncements)
+          | boolResult = (OB.addCompleted, (obj :))
+          | WC.isUnwinnable currentCompletions obj = (OB.addUnwinnable, id)
+          | otherwise = (OB.addIncomplete, id)
+
+  simplifyResult = \case
+    Left exn -> Left $ formatExn em exn
+    Right (VBool x) -> Right x
+    Right (VResult (VBool x) _env) -> Right x
+    Right val ->
+      Left $
+        T.unwords
+          [ "Non boolean value:"
+          , prettyValue val
+          , "real:"
+          , T.pack (show val)
+          ]
```

Comment:
@xsebek I have rebased atop #940. Here's where the hypothetical win check logic has moved.